crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2603,6 +2604,51 @@ impl<'db> TypeInferenceBuilder<'db> {
                     op,
                     Type::IntLiteral(i64::from(b)),
                 ),
+
+            (Type::StringLiteral(salsa_s1), Type::StringLiteral(salsa_s2)) => {
+                // Note: this relies on rust's `PartialOrd` implementation for `String` matching
+                // the behavior of Python's string comparison. Which is not exactly guaranteed.
+                let s1 = salsa_s1.value(self.db);
+                let s2 = salsa_s2.value(self.db);
```

Comment:
should this be a TODO rather than a note? I think ideally I'd want to check this and get it right. sounds like I might care about this more than @carljm though 😆

tbh I think somebody will inevitably come up with some unicode corner case that doesn't matter that much and report it as a bug. Maybe we should just avoid the problem entirely by short-circuiting if it's not an ASCII string?

```suggestion
                // For anything non-ASCII, don't even try to infer precise literal types.
                // It would be complicated, it's an unimportant edge case,
                // and Rust and Python probably do subtly different things
                // in their string methods for non-ASCII characters.
                let s1 = salsa_s1.value(self.db);
                if !s1.is_ascii() {
                    return                             Some(builtins_symbol_ty(self.db, "bool").to_instance(self.db));
                }
                let s2 = salsa_s2.value(self.db);
                if !s2.is_ascii() {
                    return                             Some(builtins_symbol_ty(self.db, "bool").to_instance(self.db));
                }
```

Reply:
I think that's better, good call!